rskj-core/src/main/java/co/rsk/cli/migration/UnitrieMigrationTool.java

```diff
@@ -0,0 +1,396 @@
+                .filter(accountAddress -> accountAddress.length == 20)
+                .collect(
+                        Collectors.toMap(accountAddress -> ByteUtil.wrap(Keccak256Helper.keccak256(accountAddress)),
+                                RskAddress::new
+                        )
+                );
+        this.addressHashes.put(ByteUtil.wrap(Keccak256Helper.keccak256(PrecompiledContracts.REMASC_ADDR.getBytes())), PrecompiledContracts.REMASC_ADDR);
+        this.addressHashes.put(ByteUtil.wrap(Keccak256Helper.keccak256(RemascTransaction.REMASC_ADDRESS.getBytes())), RemascTransaction.REMASC_ADDRESS);
+    }
+
+    public void migrate() {
+        logger.info("Migration started");
+        logger.info("Block {}", blockToMigrate.getNumber());
+        Trie migratedTrie = migrateState(blockToMigrate);
+        unitrieRepository.flush();
+
+        stateRootHandler.register(
+                blockToMigrate.getHeader(),
+                migratedTrie
+        );
+    }
+
+    private Trie migrateState(Block blockToMigrate) {
+        byte[] orchidStateRoot = blockToMigrate.getStateRoot();
+        Trie orchidAccountsTrie = orchidAccountsTrieStore.retrieve(orchidStateRoot);
+        if (!Arrays.equals(orchidStateRoot, orchidAccountsTrie.getHashOrchid(true).getBytes())) {
+            throw new IllegalStateException(String.format("Stored account state is not consistent with the expected root (%s) for block %d", Hex.toHexString(orchidStateRoot), blockToMigrate.getNumber()));
+        }
+
+        try {
+            buildPartialUnitrie(orchidAccountsTrie, unitrieRepository);
+        } catch (MissingContractStorageKeysException e) {
+            StringBuilder missingStorageKeysMessage = new StringBuilder(
+                "We have detected an inconsistency in your database and are unable to migrate it automatically.\n" +
+                "Please visit https://www.github.com/rsksmart/rskj/issues/452 for information on how to continue.\n" +
+                "Here is the data you'll need:\n"
+            );
+            for (Keccak256 entry : e.getMissingStorageKeys()) {
+                missingStorageKeysMessage.append(entry.toHexString()).append("\n");
+            }
+            logger.error(missingStorageKeysMessage.toString());
+        }
+
+        byte[] lastStateRoot = unitrieRepository.getRoot();
+        byte[] orchidMigratedStateRoot = trieConverter.getOrchidAccountTrieRoot(unitrieRepository.getMutableTrie().getTrie());
+        if (!Arrays.equals(orchidStateRoot, orchidMigratedStateRoot)) {
+            logger.error("Not matched state root");
+            logger.error("Orchid state root: {}", Hex.toHexString(orchidStateRoot));
+            logger.error("Converted Unitrie root: {}", Hex.toHexString(orchidMigratedStateRoot));
+            logger.error("Unitrie state root: {}", Hex.toHexString(lastStateRoot));
+            throw new IllegalStateException("Not matching state root");
```

Comment:
Here is the only place where you use logging instead of passing info to exceptions. I would use the same approach every place

Reply:
This exception aborts the Java process and is logged to stdout, but we need to log this for our PPA users. I improved the exception message instructing the users to read the log.